File: Gearmaker/Involute.m
//
//  Involute.m
//  Gearmaker
//
//  Created by Torsten Kammer on 27.01.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

#import "Involute.h"

@implementation Involute

- (CGPoint)pointAt:(CGFloat)angleInRad;
{
	CGFloat s = angleInRad * self.baseCircleRadius;
	
	CGPoint circumferencePoint = (CGPoint) {
		self.baseCircleRadius * cos(angleInRad),
		self.baseCircleRadius * sin(angleInRad)
	};
	
	return (CGPoint) {
		circumferencePoint.x + s*sin(angleInRad),
		circumferencePoint.y - s*cos(angleInRad)
	};
}

- (CGFloat)angleForIntersectionWithCircleRadius:(CGFloat)outerRadius;
{
	CGFloat relativeRadius = outerRadius;
	CGFloat ratio = relativeRadius / self.baseCircleRadius;
	return sqrt(ratio * ratio - 1);
}

- (CGPoint)tangentAt:(CGFloat)angleInRad;
{
	return (CGPoint) {
		cos(angleInRad),
		sin(angleInRad)
	};
}

@end

File: Gearmaker/Involute.swift
//
//  Involute.swift
//  Gearmaker
//
//  Created by Torsten Kammer on 12.11.22.
//  Copyright © 2022 Torsten Kammer. All rights reserved.
//

import Foundation

struct Involute {
    
    let baseCircleRadius: Double
    
    func point(at angleInRad: Double) -> SIMD2<Double> {
        let s = angleInRad * baseCircleRadius
        
        let circumferencePoint = CGPoint(x: baseCircleRadius * cos(angleInRad), y: baseCircleRadius * sin(angleInRad))
        
        return SIMD2<Double>(x: circumferencePoint.x + s*sin(angleInRad), y: circumferencePoint.y - s*cos(angleInRad))
    }
    
    func angleForIntersectionWithCircle(radius outerRadius: Double) -> Double {
        let relativeRadius = outerRadius
        let ratio = relativeRadius / baseCircleRadius
        return sqrt(ratio * ratio - 1) // TODO WTF?
    }
    
    func tangent(at angleInRad: Double) -> SIMD2<Double> {
        return SIMD2<Double>(x: cos(angleInRad), y: sin(angleInRad))
    }
}
